cmd/soroban-rpc/internal/db/event.go

```diff
@@ -91,14 +100,48 @@ func (eventHandler *eventHandler) InsertEvents(lcm xdr.LedgerCloseMeta) error {
 		}
 
 		query := sq.Insert(eventTableName).
-			Columns("ledger_sequence", "application_order", "contract_id", "event_type")
+			Columns("id", "contract_id", "event_type", "event_data", "ledger_close_time", "transaction_hash", "topic1", "topic2", "topic3", "topic4")
+
+		for index, e := range txEvents {
 
-		for _, e := range txEvents {
 			var contractID []byte
 			if e.Event.ContractId != nil {
 				contractID = e.Event.ContractId[:]
 			}
-			query = query.Values(lcm.LedgerSequence(), tx.Index, contractID, int(e.Event.Type))
+
+			id := Cursor{Ledger: lcm.LedgerSequence(), Tx: tx.Index, Op: 0, Event: uint32(index)}.String()
+			eventBlob, err := xdr.MarshalBase64(e)
+			if err != nil {
+				return err
+			}
+
+			v0, ok := e.Event.Body.GetV0()
```

Comment:
We don't need to store the body and the topics separately, you should just store the `data` field w/in the body

Reply:
What do you mean? We have to store topics separately in order to add index on db right?

Reply:
We do have to store the topics separately to build indices, yes (though I honestly wonder how valuable those are given that most topics are really distinct? And thus not easily sort/searchable?), but we don't need to store the entire body in a separate column since topics are part of the body, so you can just store the `data` value rather than its `body = topics + data`.

Reply:
so you meant not to duplicate topics as they are already stored separately.

overall philosophy that I have in mind for this design is to store encoded `eventXDR` first and save other metadata to reduce the latency if needed e.g. topics, contract_ids etc..

so lets say if eventXDR evolves in future and topics are now increased from 5 to 7, this design would just work fine. We won't have to do any migration or addition of any new col in DB for added topics. 

Also currently topic1-topic5 are stored on DB but we can drop some of those too later on if we see first 2 topics are good enough to reduce the fetched data. 

>(though I honestly wonder how valuable those are given that most topics are really distinct? And thus not easily sort/searchable?)

hmm I mean it does help with requested invalid topic tho. but yeah without indexing it is at `~200-250 ms `when benchmarked against `30 millions events.` @Shaptic

Reply:
Good points all around, duplication of the structure seems worth it to avoid later complications :+1: